samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java

```diff
@@ -128,6 +150,17 @@ public final void onMessage(M message, MessageCollector collector, TaskCoordinat
     this.handleMessageNs.update(endNs - startNs);
 
     results.forEach(rm -> this.registeredOperators.forEach(op -> op.onMessage(rm, collector, coordinator)));
+
+    InitableFunction transformFn = getOperatorSpec().getTransformFn();
+    if (transformFn != null && transformFn instanceof WatermarkFunction) {
+      // check whether there is new watermark emitted from the user function
+      long outputWm = ((WatermarkFunction) transformFn).getOutputWatermark();
+      if (outputWatermark < outputWm) {
```

Comment:
I'd probably add a warn to the log if outputWatermark > outputWm. It is OK for outputWatermark == outputWm and the current behavior is correct.